neuralprophet/conformal.py

```diff
@@ -48,34 +51,50 @@ def predict(self, df: pd.DataFrame, df_cal: pd.DataFrame) -> pd.DataFrame:
                     test dataframe with uncertainty prediction intervals
 
         """
-        # conformalize
-        self.noncon_scores = self._get_nonconformity_scores(df_cal)
-        self.q_hat = self._get_q_hat(df_cal)
-        df["qhat1"] = self.q_hat
-        if self.method == "naive":
-            df["yhat1 - qhat1"] = df["yhat1"] - self.q_hat
-            df["yhat1 + qhat1"] = df["yhat1"] + self.q_hat
-        elif self.method == "cqr":
-            quantile_hi = str(max(self.quantiles) * 100)
-            quantile_lo = str(min(self.quantiles) * 100)
-            df[f"yhat1 {quantile_hi}% - qhat1"] = df[f"yhat1 {quantile_hi}%"] - self.q_hat
-            df[f"yhat1 {quantile_hi}% + qhat1"] = df[f"yhat1 {quantile_hi}%"] + self.q_hat
-            df[f"yhat1 {quantile_lo}% - qhat1"] = df[f"yhat1 {quantile_lo}%"] - self.q_hat
-            df[f"yhat1 {quantile_lo}% + qhat1"] = df[f"yhat1 {quantile_lo}%"] + self.q_hat
-        else:
-            raise ValueError(
-                f"Unknown conformal prediction method '{self.method}'. Please input either 'naive' or 'cqr'."
-            )
+        self.q_hats = []
+        for step_number in range(1, self.n_forecasts + 1):
+            # conformalize
+            noncon_scores = self._get_nonconformity_scores(df_cal, step_number)
+            q_hat = self._get_q_hat(df_cal, noncon_scores)
+            df[f"qhat{step_number}"] = q_hat
+            if self.method == "naive":
+                df[f"yhat{step_number} - qhat{step_number}"] = df[f"yhat{step_number}"] - q_hat
+                df[f"yhat{step_number} + qhat{step_number}"] = df[f"yhat{step_number}"] + q_hat
+            elif self.method == "cqr":
+                quantile_hi = str(max(self.quantiles) * 100)
+                quantile_lo = str(min(self.quantiles) * 100)
+                df[f"yhat{step_number} {quantile_hi}% - qhat{step_number}"] = (
+                    df[f"yhat{step_number} {quantile_hi}%"] - q_hat
+                )
+                df[f"yhat{step_number} {quantile_hi}% + qhat{step_number}"] = (
+                    df[f"yhat{step_number} {quantile_hi}%"] + q_hat
+                )
+                df[f"yhat{step_number} {quantile_lo}% - qhat{step_number}"] = (
+                    df[f"yhat{step_number} {quantile_lo}%"] - q_hat
+                )
+                df[f"yhat{step_number} {quantile_lo}% + qhat{step_number}"] = (
+                    df[f"yhat{step_number} {quantile_lo}%"] + q_hat
+                )
+            else:
+                raise ValueError(
+                    f"Unknown conformal prediction method '{self.method}'. Please input either 'naive' or 'cqr'."
+                )
+            if step_number == 1:
+                # save nonconformity scores of the first timestep
```

Comment:
Why are these saved for the first step (only)?

Reply:
The nonconformity scores are saved in order to be inputted into `plot_nonconformity_scores()` , where it plots:
![Naive One-Sided Interval Width with q](https://user-images.githubusercontent.com/13785749/212239966-6dc5087a-f000-4bc8-8870-ffb47ff9f57b.png)

The nonconformity score is the score (blue line).

This plot is only called when n_forecasts ==1. For models with n_forecasts >1, it will call the `plot_interval_width_per_timestep()`  instead (more details under your 3rd question), which doesn't require nonconformity scores. Therefore, only the first step needs to be saved.

Reply:
Thank you for the explanation!
From a UI-perspective it's not ideal to have the plot change quite so drastically depending on an external factor, but I think it is OK in this case as this is more of a research/diagnostic plot.